### PR TITLE
ci(nss): handle system-, cached-, and to-be-build-NSS

### PR DESCRIPTION
**Pull request by mxinden**
When the system NSS version is sufficient:

https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/.github/actions/nss/action.yml#L31

Then one must not set the environment variables pointing to a local copy:

https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/.github/actions/nss/action.yml#L177-L190

Fixes https://github.com/mozilla/neqo/issues/2141.

### CODE REVIEW

**Comment by larseggert** [comment]
Is this fixing a bug?

When a prebuilt NSS is downloaded from the GitHub cache, `BUILD_NSS` is set to 0 but the environment still needs to be set up so it is found by the neqo build process.

**Comment by mxinden** [comment]
> coalesce_acked_from_zero 10+1 entries: 💚 Performance has improved.
> 
>        time:   [116.48 ns 116.99 ns 117.58 ns]
>        change: [-39.843% -35.572% -33.050%] (p = 0.00 < 0.05)
> 
> Found 15 outliers among 100 measurements (15.00%)

Ufff, a 35% performance improvement on an unrelated change :thinking:

**Comment by mxinden** [comment]
> Is this fixing a bug?

Yes, this fixes https://github.com/mozilla/neqo/issues/2141.

> When a prebuilt NSS is downloaded from the GitHub cache, `BUILD_NSS` is set to 0 but the environment still needs to be set up so it is found by the neqo build process.

When `NSS_DIR` isn't set, `pkg-config` is used to find the system-wide installation of NSS:

https://github.com/mozilla/neqo/blob/55e3a9363c28632dfb29ce91c7712cab1f6a58da/neqo-crypto/build.rs#L428-L434

Thus I don't think the environment variables need to be set.

@larseggert am I missing something?

**Comment by larseggert** [comment]
You are right that if `NSS_DIR` isn't set, we try and use the system NSS.

But if we need to use an NSS that was build from source, we need to set the environment variables, so that neqo can link against the NSS that was build by (or cache-restored by) the GitHub action.

With your change, is it still using the prebuilt-from-source NSS, or is it building it again as part of neqo?

**Comment by mxinden** [comment]
I think I understand know. Thanks @larseggert. I missed that `BUILD_NSS` is overwritten on cache hit:

https://github.com/mozilla/neqo/blob/a484f64f64da94376f05f11cf68b8894cab18bd0/.github/actions/nss/action.yml#L124-L125

Let me take another look.

File: .github/actions/nss/action.yml [hunk 1]
```diff
@@ -177,17 +178,25 @@ runs:
...
+          NSS_OUT="$NSS_DIR/../dist/$NSS_TARGET"
+          echo "LD_LIBRARY_PATH=$NSS_OUT/lib" >> "$GITHUB_ENV"
+          echo "DYLD_FALLBACK_LIBRARY_PATH=$NSS_OUT/lib" >> "$GITHUB_ENV"
+          echo "$NSS_OUT/lib" >> "$GITHUB_PATH"
+          echo "NSS_DIR=$NSS_DIR" >> "$GITHUB_ENV"
+          echo "NSS_PREBUILT=1" >> "$GITHUB_ENV"
+        fi
       env:
         NSS_DIR: ${{ github.workspace }}/nss
-        NSPR_DIR: ${{ github.workspace }}/nspr
```
Comment (mxinden) [inline .github/actions/nss/action.yml:190]: As far as I can tell, this environment variable is never used.

File: .github/actions/nss/action.yml [hunk 2]
```diff
@@ -86,21 +86,21 @@ runs:
         fi
 
     - name: Checkout NSS
-      if: env.BUILD_NSS == '1'
+      if: env.USE_SYSTEM == '0'
```
Comment (martinthomson) [inline .github/actions/nss/action.yml:89]: USE_SYSTEM_NSS !

Comment (mxinden) [reply .github/actions/nss/action.yml:89]: Uff, good catch. Thank you @martinthomson! Will be fixed in https://github.com/mozilla/neqo/pull/2145.

**Pull request merged**